presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -71,418 +86,748 @@
-        try {
-            MaterializedViewInformationExtractor materializedViewInformationExtractor = new MaterializedViewInformationExtractor();
-            materializedViewInformationExtractor.process(materializedViewQuery);
-            materializedViewInfo = materializedViewInformationExtractor.getMaterializedViewInfo();
-            return process(node);
-        }
-        catch (Exception ex) {
-            logger.warn("Failed to rewrite query with materialized view with following exception: %s", ex.getMessage());
+        if (!node.getFrom().isPresent()) {
+            log.debug("Visited QuerySpecification had no from field: \n%s", node);
             return node;
         }
+        // If a query specification has a Table as a source, it can potentially be rewritten,
+        // so hand control over to LeafSubqueryVisitor via rewriteQuerySpecificationIfCompatible
+        Relation from = node.getFrom().get();
+        if (from instanceof Table) {
+            return rewriteQuerySpecificationIfCompatible(node, (Table) from);
+        }
+
+        if (from instanceof AliasedRelation
+                && ((AliasedRelation) from).getRelation() instanceof Table) {
+            AliasedRelation aliasedRelation = (AliasedRelation) from;
+            return rewriteQuerySpecificationIfCompatible(node, (Table) aliasedRelation.getRelation());
+        }
+
+        Relation newFrom = processSameType(from);
+        if (from == newFrom) {
+            return node;
+        }
+
+        if (node.getLocation().isPresent()) {
+            return new QuerySpecification(
+                    node.getLocation().get(),
+                    node.getSelect(),
+                    Optional.of(newFrom),
+                    node.getWhere(),
+                    node.getGroupBy(),
+                    node.getHaving(),
+                    node.getOrderBy(),
+                    node.getOffset(),
+                    node.getLimit());
+        }
+
+        return new QuerySpecification(
+                node.getSelect(),
+                Optional.of(newFrom),
+                node.getWhere(),
+                node.getGroupBy(),
+                node.getHaving(),
+                node.getOrderBy(),
+                node.getOffset(),
+                node.getLimit());
     }
 
     @Override
     protected Node visitNode(Node node, Void context)
     {
-        for (Node child : node.getChildren()) {
-            process(child, context);
-        }
         return node;
     }
 
     @Override
     protected Node visitQuery(Query node, Void context)
     {
+        QueryBody newQueryBody = processSameType(node.getQueryBody());
+        Optional<With> newWith = node.getWith().map(this::processSameType);
+        boolean withSame = !node.getWith().isPresent() || (node.getWith().get()) == newWith.get();
+        if (withSame && (node.getQueryBody() == newQueryBody)) {
```

Comment:
Parentheses are redundant
```suggestion
if (withSame && node.getQueryBody() == newQueryBody) {
```